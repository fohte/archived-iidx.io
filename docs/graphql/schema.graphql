enum ClearLamp {
  """
  Assist Clear
  """
  ASSIST

  """
  Easy Clear
  """
  EASY

  """
  EX Hard Clear
  """
  EX_HARD

  """
  Failed
  """
  FAILED

  """
  FullCombo Clear
  """
  FULL_COMBO

  """
  Hard Clear
  """
  HARD

  """
  Clear
  """
  NORMAL
}

"""
Autogenerated return type of CreateUser
"""
type CreateUserPayload {
  user: User!
}

enum Difficulty {
  """
  Another
  """
  ANOTHER

  """
  Hyper
  """
  HYPER

  """
  Normal
  """
  NORMAL
}

enum Grade {
  """
  A
  """
  A

  """
  AA
  """
  AA

  """
  AAA
  """
  AAA

  """
  B
  """
  B

  """
  C
  """
  C

  """
  D
  """
  D

  """
  E
  """
  E

  """
  F
  """
  F
}

type Map {
  bestResult(username: String!): Result
  difficulty: Difficulty!
  id: ID!
  level: Int!
  maxBpm: Int!
  minBpm: Int!
  music: Music!
  numNotes: Int!
  playStyle: PlayStyle!
}

type Music {
  artist: String!
  genre: String!
  id: ID!
  leggendaria: Boolean!
  map(difficulty: Difficulty!, playStyle: PlayStyle!): Map
  maps: [Map!]!
  series: Int!
  subTitle: String!
  textageUid: String!
  title: String!
}

type Mutation {
  createUser(displayName: String, username: String!): CreateUserPayload
  registerResultsWithCSV(csv: String!, playStyle: PlayStyle!): RegisterResultsWithCSVPayload
}

enum PlayStyle {
  """
  DP
  """
  DP

  """
  SP
  """
  SP
}

type Query {
  """
  Find maps.
  """
  maps: [Map!]

  """
  Find a music.
  """
  music(id: ID!): Music

  """
  Find musics.
  """
  musics: [Music!]

  """
  Find a user by name.
  """
  user(name: String!): User

  """
  The currently authenticated user.
  """
  viewer: User
}

"""
Autogenerated return type of RegisterResultsWithCSV
"""
type RegisterResultsWithCSVPayload {
  success: Boolean!
}

type Result {
  clearLamp: ClearLamp!
  grade: Grade!
  id: ID!
  missCount: Int!
  score: Int!
}

type User {
  id: ID!
  name: String!
  profile: UserProfile!
}

type UserProfile {
  displayName: String!
  id: ID!
}