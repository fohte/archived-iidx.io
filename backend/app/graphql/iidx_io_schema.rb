# frozen_string_literal: true

class IIDXIOSchema < GraphQL::Schema
  mutation(Types::MutationType)
  query(Types::QueryType)
end
