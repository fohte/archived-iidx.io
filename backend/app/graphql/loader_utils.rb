# frozen_string_literal: true

module LoaderUtils
  class << self
    def find_by(*args, **kwargs, &block)
      promise = load_promise(*args, **kwargs)

      if block_given?
        promise.then(&block)
      else
        promise.then { |record| record }
      end
    end

    def find_by!(model, column_values, **kwargs)
      promise = load_promise(model, column_values, **kwargs)

      promise.then do |record|
        raise IIDXIO::GraphQL::NotFoundError, "Couldn't find #{model.name} with #{column_values.to_json}" if record.nil?

        if block_given?
          yield(record)
        else
          record
        end
      end
    end

    private

    def load_promise(model, column_values, scope: nil)
      Loaders::RecordLoader
        .for(model, columns: column_values.keys, scope: scope)
        .load(column_values.values)
    end
  end
end
