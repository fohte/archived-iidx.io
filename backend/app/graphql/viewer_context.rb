# frozen_string_literal: true

class ViewerContext < GraphQL::Query::Context
  def viewer
    self[:viewer]
  end

  def firebase_uid
    self[:firebase_uid]
  end
end
