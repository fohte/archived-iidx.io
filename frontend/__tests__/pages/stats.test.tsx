import { MockedProvider, MockedResponse } from '@apollo/react-testing'
import { GraphQLError } from 'graphql'
import * as React from 'react'
import { mount } from 'enzyme'
import * as TestUtils from 'react-dom/test-utils'

import Stats, { Props } from '@pages/stats'
import {
  PlayStyle,
  FetchStatsQuery,
  FetchStatsDocument,
  FetchStatsQueryVariables,
} from '@app/queries'

const wait = (amount = 0) => new Promise(resolve => setTimeout(resolve, amount))

function act<T>(callback: () => T): T {
  let ret: T

  TestUtils.act(() => {
    ret = callback()
  })

  // @ts-ignore "Variable is used before being assigned" エラーを無視する
  // TestUtils.act で ret 変数には確実に値が代入されているため
  return ret
}

const baseProps: Required<Props> = {
  playStyle: PlayStyle.Sp,
  screenName: 'user',
}

type User = NonNullable<FetchStatsQuery['user']>
type LevelGradeMatrix = User['countByEachLevelAndGrade']

const createRequest = () => {
  const variables: FetchStatsQueryVariables = {
    playStyle: baseProps.playStyle,
    username: baseProps.screenName,
  }

  return {
    query: FetchStatsDocument,
    variables,
  }
}

const createMock = (levelGradeMatrix: LevelGradeMatrix): MockedResponse[] => {
  const request = createRequest()

  const data: FetchStatsQuery = {
    user: {
      countByEachLevelAndGrade: levelGradeMatrix,
    },
  }

  return [{ request, result: { data } }]
}

describe('/map', () => {
  it('renders correctly', async () => {
    const wrapper = act(() =>
      mount(
        <MockedProvider mocks={createMock([])} addTypename={false}>
          <Stats {...baseProps} />
        </MockedProvider>,
      ),
    )

    await TestUtils.act(async () => {
      // wait for response of the query
      await wait(0)
    })

    expect(wrapper.getDOMNode()).toMatchSnapshot()
  })

  it('renders correctly when loading', () => {
    const wrapper = mount(
      <MockedProvider mocks={createMock([])} addTypename={false}>
        <Stats {...baseProps} />
      </MockedProvider>,
    )

    expect(wrapper.getDOMNode()).toMatchSnapshot()
  })

  it('renders correctly if there are no required arguments', () => {
    const wrapper = mount(<Stats />)

    expect(wrapper.getDOMNode()).toMatchSnapshot()
  })

  it('renders correctly if there are graphql errors', async () => {
    const mocks: MockedResponse[] = [
      {
        request: createRequest(),
        result: { errors: [new GraphQLError('some error')] },
      },
    ]

    const wrapper = act(() =>
      mount(
        <MockedProvider mocks={mocks} addTypename={false}>
          <Stats {...baseProps} />
        </MockedProvider>,
      ),
    )

    await TestUtils.act(async () => {
      // wait for response of the query
      await wait(0)
    })

    expect(wrapper.getDOMNode()).toMatchSnapshot()
  })
})
