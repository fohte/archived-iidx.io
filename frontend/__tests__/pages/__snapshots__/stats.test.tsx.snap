// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/map renders correctly 1`] = `
<div
  class="box"
>
  <nav
    class="headerNav"
  >
    <div
      class="container "
    >
      <div
        class="flexContainer"
      >
        <a
          href="/"
        >
          <div
            class="item"
          >
            <span>
              iidx.io
            </span>
          </div>
        </a>
        <div
          class="item"
        >
          loading
        </div>
      </div>
    </div>
  </nav>
  <div
    class="content"
  >
    <div
      class="container "
    >
      <div
        class="user-profile-header-wrapper"
      >
        <div
          class="user-profile-header"
        >
          <div
            class="user-profile-content"
          >
            <div
              class="user-avatar"
            />
            <h2>
              @
              user
            </h2>
          </div>
        </div>
        <div
          class="tabs-wrapper"
        >
          <ul
            class="tabs"
          >
            <li
              class="tab-item"
            >
              <a
                href="/@user/sp"
              >
                <span
                  class="icon-text"
                >
                  Overview
                </span>
              </a>
            </li>
            <li
              class="tab-item"
            >
              <a
                href="/@user/sp/musics"
              >
                <span
                  class="icon-text"
                >
                  Musics
                </span>
              </a>
            </li>
            <li
              class="tab-item active"
            >
              <a
                href="/@user/sp/stats"
              >
                <span
                  class="icon-text"
                >
                  Stats
                </span>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      class="container "
    >
      <div
        class="box "
      >
        <div
          class="box-header"
        >
          <h2>
            Grades
          </h2>
        </div>
        <div
          class="table-shadow"
        >
          <div
            class="table-wrapper"
          >
            <table
              class="table"
            >
              <thead>
                <tr>
                  <th
                    class="table-head-column"
                  >
                    <div />
                  </th>
                  <th>
                    AAA
                  </th>
                  <th>
                    AA
                  </th>
                  <th>
                    A
                  </th>
                  <th>
                    B
                  </th>
                  <th>
                    C
                  </th>
                  <th>
                    D
                  </th>
                  <th>
                    E
                  </th>
                  <th>
                    F
                  </th>
                  <th>
                    NO PLAY
                  </th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  <th
                    class="table-head-column"
                  >
                    <div>
                      ☆12
                    </div>
                  </th>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                </tr>
                <tr>
                  <th
                    class="table-head-column"
                  >
                    <div>
                      ☆11
                    </div>
                  </th>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                </tr>
                <tr>
                  <th
                    class="table-head-column"
                  >
                    <div>
                      ☆10
                    </div>
                  </th>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                </tr>
                <tr>
                  <th
                    class="table-head-column"
                  >
                    <div>
                      ☆9
                    </div>
                  </th>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                </tr>
                <tr>
                  <th
                    class="table-head-column"
                  >
                    <div>
                      ☆8
                    </div>
                  </th>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                </tr>
                <tr>
                  <th
                    class="table-head-column"
                  >
                    <div>
                      ☆7
                    </div>
                  </th>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                </tr>
                <tr>
                  <th
                    class="table-head-column"
                  >
                    <div>
                      ☆6
                    </div>
                  </th>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                </tr>
                <tr>
                  <th
                    class="table-head-column"
                  >
                    <div>
                      ☆5
                    </div>
                  </th>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                </tr>
                <tr>
                  <th
                    class="table-head-column"
                  >
                    <div>
                      ☆4
                    </div>
                  </th>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                </tr>
                <tr>
                  <th
                    class="table-head-column"
                  >
                    <div>
                      ☆3
                    </div>
                  </th>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                </tr>
                <tr>
                  <th
                    class="table-head-column"
                  >
                    <div>
                      ☆2
                    </div>
                  </th>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                </tr>
                <tr>
                  <th
                    class="table-head-column"
                  >
                    <div>
                      ☆1
                    </div>
                  </th>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                </tr>
              </tbody>
              <tfoot>
                <tr>
                  <th
                    class="table-head-column"
                  >
                    <div>
                      Total
                    </div>
                  </th>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                  <td>
                    0
                  </td>
                </tr>
              </tfoot>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
  <footer
    class="box"
  >
    <div
      class="container "
    >
      <div
        class="inner"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-copyright fa-w-16 "
          data-icon="copyright"
          data-prefix="far"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8C119.033 8 8 119.033 8 256s111.033 248 248 248 248-111.033 248-248S392.967 8 256 8zm0 448c-110.532 0-200-89.451-200-200 0-110.531 89.451-200 200-200 110.532 0 200 89.451 200 200 0 110.532-89.451 200-200 200zm107.351-101.064c-9.614 9.712-45.53 41.396-104.065 41.396-82.43 0-140.484-61.425-140.484-141.567 0-79.152 60.275-139.401 139.762-139.401 55.531 0 88.738 26.62 97.593 34.779a11.965 11.965 0 0 1 1.936 15.322l-18.155 28.113c-3.841 5.95-11.966 7.282-17.499 2.921-8.595-6.776-31.814-22.538-61.708-22.538-48.303 0-77.916 35.33-77.916 80.082 0 41.589 26.888 83.692 78.277 83.692 32.657 0 56.843-19.039 65.726-27.225 5.27-4.857 13.596-4.039 17.82 1.738l19.865 27.17a11.947 11.947 0 0 1-1.152 15.518z"
            fill="currentColor"
          />
        </svg>
         
        2018
         iidx.io |
         
        test
      </div>
    </div>
  </footer>
</div>
`;

exports[`/map renders correctly if there are graphql errors 1`] = `null`;

exports[`/map renders correctly if there are no required arguments 1`] = `
<div
  style="color: rgb(0, 0, 0); background: rgb(255, 255, 255); font-family: -apple-system, BlinkMacSystemFont, Roboto, \\"Segoe UI\\", \\"Fira Sans\\", Avenir, \\"Helvetica Neue\\", \\"Lucida Grande\\", sans-serif; height: 100vh; text-align: center; display: flex; flex-direction: column; align-items: center; justify-content: center;"
>
  <div>
    <style>
      body { margin: 0 }
    </style>
    <h1
      style="display: inline-block; margin: 0px 20px 0px 0px; padding: 10px 23px 10px 0px; font-size: 24px; font-weight: 500; vertical-align: top;"
    >
      404
    </h1>
    <div
      style="display: inline-block; text-align: left; line-height: 49px; height: 49px; vertical-align: middle;"
    >
      <h2
        style="font-size: 14px; font-weight: normal; line-height: inherit; margin: 0px; padding: 0px;"
      >
        This page could not be found
        .
      </h2>
    </div>
  </div>
</div>
`;

exports[`/map renders correctly when loading 1`] = `
<div
  class="box"
>
  <nav
    class="headerNav"
  >
    <div
      class="container "
    >
      <div
        class="flexContainer"
      >
        <a
          href="/"
        >
          <div
            class="item"
          >
            <span>
              iidx.io
            </span>
          </div>
        </a>
        <div
          class="item"
        >
          loading
        </div>
      </div>
    </div>
  </nav>
  <div
    class="content"
  >
    <div
      class="container "
    >
      <div
        class="user-profile-header-wrapper"
      >
        <div
          class="user-profile-header"
        >
          <div
            class="user-profile-content"
          >
            <div
              class="user-avatar"
            />
            <h2>
              @
              user
            </h2>
          </div>
        </div>
        <div
          class="tabs-wrapper"
        >
          <ul
            class="tabs"
          >
            <li
              class="tab-item"
            >
              <a
                href="/@user/sp"
              >
                <span
                  class="icon-text"
                >
                  Overview
                </span>
              </a>
            </li>
            <li
              class="tab-item"
            >
              <a
                href="/@user/sp/musics"
              >
                <span
                  class="icon-text"
                >
                  Musics
                </span>
              </a>
            </li>
            <li
              class="tab-item active"
            >
              <a
                href="/@user/sp/stats"
              >
                <span
                  class="icon-text"
                >
                  Stats
                </span>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      class="container "
    />
  </div>
  <footer
    class="box"
  >
    <div
      class="container "
    >
      <div
        class="inner"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-copyright fa-w-16 "
          data-icon="copyright"
          data-prefix="far"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8C119.033 8 8 119.033 8 256s111.033 248 248 248 248-111.033 248-248S392.967 8 256 8zm0 448c-110.532 0-200-89.451-200-200 0-110.531 89.451-200 200-200 110.532 0 200 89.451 200 200 0 110.532-89.451 200-200 200zm107.351-101.064c-9.614 9.712-45.53 41.396-104.065 41.396-82.43 0-140.484-61.425-140.484-141.567 0-79.152 60.275-139.401 139.762-139.401 55.531 0 88.738 26.62 97.593 34.779a11.965 11.965 0 0 1 1.936 15.322l-18.155 28.113c-3.841 5.95-11.966 7.282-17.499 2.921-8.595-6.776-31.814-22.538-61.708-22.538-48.303 0-77.916 35.33-77.916 80.082 0 41.589 26.888 83.692 78.277 83.692 32.657 0 56.843-19.039 65.726-27.225 5.27-4.857 13.596-4.039 17.82 1.738l19.865 27.17a11.947 11.947 0 0 1-1.152 15.518z"
            fill="currentColor"
          />
        </svg>
         
        2018
         iidx.io |
         
        test
      </div>
    </div>
  </footer>
</div>
`;
